### ISSUE TITLE
New Page asXml() truncated

### ISSUE THREAD

**Issue by lrpowell** [HtmlUnit/htmlunit#337]
I believe I am launching into the new page that will display the chart on the browser.  But the html returned is not complete no matter how long I wait (the chart does take a while when rendered in browser).

```
		HtmlPage newPage = anchorElement.click(false, false, false, true, true, false);
		System.out.println("override click ");
		int count2 = 0;
		boolean foundLoading = false;
	        String LOADING_STR = "LOADING...";
		while (!foundLoading && count2++ < 200) {
			webClient.waitForBackgroundJavaScriptStartingBefore(2000);
			foundLoading = newPage.asXml().contains("Graph Interval")
				&& newPage.asXml().contains(LOADING_STR);
			if (foundLoading) {
				outputToFile("withchart.html", newPage.asXml());
				// outputToFile("chartAsNText.txt", newPage.asNormalizedText()); // empty
			}
			if (count2 % 100 == 0)
				System.out.println("LOADING count2:" + count2);
		}
		// now waiting for loading to go away
		count2 = 0;
		while (foundLoading && count2 < 10000) {
			if (count2++ % 100 == 0)
				System.out.println("waiting for LOADING to go away - count2:" + count2);
			webClient.waitForBackgroundJavaScript(2000);
			foundLoading = newPage.asXml().contains(LOADING_STR);
			if (!foundLoading) {
				outputToFile("chartLoaded.html", newPage.asXml());
				outputToFile("loadedChartAsNText.txt", newPage.asNormalizedText());
			}
		}
		outputToFile("chart_loaded.html", newPage.asXml());

```

[chart_loaded.html.txt](https://github.com/HtmlUnit/htmlunit/files/6281374/chart_loaded.html.txt)

The last line is:

 `<script src="/Navigator/assets/pt/dojoChart/DojoGraphLauncher.js" type="text/javascript" graphinfo="{&quot;packageId&quot;:&quot;IBM_MN&quot;,&quot;perspectiveId&quot;:&quot;P_SM_CPU_UTIL_AVG&quot;,&quot;perspectiveName&quot;:&quot;CPU Utilization (Average)&quot;,&quot;collectionLib&quot;:&quot;QPFRDATA&quot;,&quot;collectionName&quot;:&quot;Q098000004&quot;,&quot;collectionType&quot;:&quot;*CSFILE&quot;,&quot;collectionFormat&quot;:&quot;*CSFILE&quot;,&quot;fileLevel&quot;:&quot;48&quot;,&quot;startTime&quot;:&quot;Apr 8, 2021, 12:00:04 AM&quot;,&quot;endTime&quot;:&quot;ONGOING&quot;,&quot;systemName&quot;:&quot;Z1432BP3&quot;,&quot;release&quot;:&quot;V7R3M0&quot;,&quot;targetSystem&quot;:&quot;z1432bp3.rch.stglabs.ibm.com&quot;}" contextmri="{&quot;context&quot;:&quot;Context&quot;,&quot;actions&quot;:&quot;Actions&quot;,&quot;collectionName&quot;:&quot;Collection Name:&quot;,&quot;collectionLib&quot;:&quot;Collection Library:&quot;,&quot;fileLevel&quot;:&quot;File Level:&quot;,&quot;startTime&quot;:&quot;Start Time:&quot;,&quot;endTime&quot;:&quot;End Time:&quot;,&quot;systemName&quot;:&quot;System Name:&quot;,&quot;release&quot;:&quot;Release:&quot;,&quot;EMPTY_RESULT_SET&quot;:&quot;Query returned empty result set. No data is available to display in chart or table.&quot;,&quot;EMPTY_RESULT_SET2&quot;:&quot;Query returned empty result set. There is no data available to display in chart or table.&quot;,&quot;refresh&quot;:&quot;Refresh&quot;,&quot;reset&quot;:&quot;Reset&quot;,&quot;CONTEXT_FROM&quot;:&quot;From&quot;,&quot;CONTEXT_TO&quot;:&quot;To&quot;,&quot;loading&quot;:&quot;LOADING....&quot;,&quot;NO_RESULTS_FROM_SERVER&quot;:&quot;No results returned from server&quot;,&quot;HELP_ICON_TEXT&quot;:&quot;Click for help&quot;,&quot;collectionType&quot;:&quot;Collection Type:&quot;,&quot;graphInterval&quot;:&quot;Graph Interval:&quot;,&quot;INTNUM&quot;:&quot;Interval number&quot;,&quot;T_TABLE_VIEW&quot;:&quot;Table View&quot;,&quot;T_TABLE_CHART_VIEW&quot;:&quot;Chart &amp; Table View&quot;,&quot;T_CHART_VIEW&quot;:&quot;Chart View&quot;,&quot;T_SHOW_SQL&quot;:&quot;Show SQL&quot;,&quot;1min&quot;:&quot;1 minute&quot;,&quot;5min&quot;:&quot;5 minute&quot;,&quot;15min&quot;:&quot;15 minute&quot;,&quot;30min&quot;:&quot;30 minute&quot;,&quot;1hour&quot;:&quot;1 hour&quot;,&quot;6hour&quot;:&quot;6 hour&quot;,&quot;12hour&quot;:&quot;12 hour&quot;,&quot;1day&quot;:&quot;1 day&quot;,&quot;viewAllPoints&quot;:&quot;View all points&quot;,&quot;E_ERROR&quot;:&quot;E_ERROR&quot;,&quot;QUERY_PARSE_ERROR&quot;:&quot;The query you entered can not be understood by this application. Please correct and try again.&quot;,&quot;MODIFY_SQL&quot;:&quot;Modify SQL&quot;,&quot;T_HELP&quot;:&quot;Help&quot;,&quot;OK&quot;:&quot;OK&quot;,&quot;CANCEL&quot;:&quot;CANCEL&quot;,&quot;SYSTEM_INFOR&quot;:&quot;System Information&quot;,&quot;CSFILEFORMAT&quot;:&quot;Collection Services File Based Collection&quot;,&quot;QAPMCONF_VIEW&quot;:&quot;QAPMCO`

How can I debug to find out why it stops at this point?  I have the output at ALL and don't see anything significant other than items I have reported earlier.
[debug output htmlunit.txt](https://github.com/HtmlUnit/htmlunit/files/6281454/debug.output.htmlunit.txt)

=== overview of using HtmlUnit for this project ===

I am starting to lose hope that HtmlUnit is going to work with my project.  I have tried several paths and keep getting stopped with parts that do not work for my project.  

issue #335 - cannot view text of page (asNormalizedText() is blank) to look for status messages such as "LOADING...".  Should be able to circumvent this by identifying something that will only be in the "asXml()" output when the loading portion of the chart is completed.

issue #324 - cannot click on link in table with Typescript Angular code.  Have circumvented this with using the menu option and selecting the drop down menu list item

issue #318 - errors that may or may not be causing problems with the chart rendering

issue #310 - seems insignificant to my project

issue #305 - cannot export the chart to a downloadable file.  Working through this another direction and I have some hope that using this new way could avoid this problem (?).  This will be the biggest blocker if I hit it again with the new direction.

**Comment by lrpowell** [comment]
The final line is 2985 characters (with spaces).  Is there a string limit reached?

**Comment by rbri** [comment]
> I am starting to lose hope that HtmlUnit is going to work with my project. I have tried several paths and keep getting stopped with parts that do not work for my project.

Hi Lora,
sorry for not being super responsive during the last weeks. But i have to solve some issues for my real job - there is no so much time left for working on all this stuff.
And maybe you are now reaching the 'Trough of Disillusionment' (https://en.wikipedia.org/wiki/Hype_cycle) [just kidding]

As always i will try to work on your (and all the other) issues on a best effort base (or maybe you have to think about sponsoring?)

**Comment by rbri** [comment]
Back to this issue.

There is no limit - as fas a i remember. The only limit is related to the way responses are saved in the WebClient - usually the response is saved in memory but if the limit is reached the response goes into a temp file.
You can change the value by using the WebClientOption#setMaxInMemory(numberOfByotes). The default value is 500 * 1024.

Please change the value and check if this changes something.

**Comment by rbri** [comment]
And another point - at least some chart libraries are using animation based on Window.requestAnimationFrame() to make some nice looking charts that 'evolve'. This might be a reason why your page does not finish.

For the last release (done yesterday) i made some documentation how to deal with this - https://htmlunit.sourceforge.io/details.html. It might be worth a try to update your wait code with this in mind.

**Comment by lrpowell** [comment]
Bring on the slope of enlightenment!
What does sponsoring entail?

**Comment by lrpowell** [comment]
Thanks for the ideas.  Does the maxInMemory do multiple things besides determine when output is written to a file?  The wording with the "history size" is confusing to me.  I tried setting MaxInMemory to 0 (Sets the maximum bytes to have in memory, after which the content is saved to a file. Set this to zero or -1 to deactivate the history at all) as well as Integer.MAX_VALUE.  Also tried setting history size to 0 and page cache limit to 0.

I don't have any animated frames, but put in code to check that there are none.

> 1-MaxinMem:512000
> 2- hist size limit:50
> 3-pageCache limit:2147483647
> maxinMem:0
> HistSizeLimit:0
> histPageCacheLim:0
> processing pending frame:0

No changes in results at this time, it is still cut off at exactly the same characters.

**Comment by rbri** [comment]
> Bring on the slope of enlightenment! What does sponsoring entail?

The idea is to be able to work one day per month (or maybe more if there is more sponsoring) for the sponsor - see

**Comment by rbri** [comment]
Strange:

what happens if you call asXml() for part of the page e.g. only for the body node?

**Comment by rbri** [comment]
Oh, looks like your page is really huge and includes a lot of javascript code.
In fact internally a StringBuilder is use and there is a limit (https://stackoverflow.com/questions/38067717/how-many-characters-can-a-java-stringbuilder-hold).
Maybe you reach that limit?

**Comment by lrpowell** [comment]
I will give that a try.  Did you insert a link or something that didn't come out above in the note about sponsering?

**Comment by lrpowell** [comment]
I will agree to sponsoring the htmlunit project if we get it to work for my project!  I can do some now to help get it to work.  How would I start?  I'll look around it is probably documented somewhere...

**Comment by rbri** [comment]
Yes, i set up github sponsoring - check the project hompage or the github page.

**Comment by rbri** [comment]
> what happens if you call asXml() for part of the page e.g. only for the body node?

Hi Lora, did you gave up?

**Comment by lrpowell** [comment]
No, I haven't given up.  But I did put it on the shelf for a while and now of course busy with other things.  I was hoping that one of these problems could get resolved before spending more time on it.  I will try what you asked this week.
